bpfd/src/bpf.rs

```diff
@@ -41,18 +41,145 @@ const MAPS_MODE: u32 = 0o0660;
+                } else {
+                    false
+                }
+            })
+            .collect::<HashMap<_, _>>()
+            .len()
+    }
+
+    fn collect_programs(
+        &mut self,
+        program_type: ProgramType,
+        if_index: Option<u32>,
+        direction: Option<Direction>,
+    ) -> Vec<(&Uuid, &mut Program)> {
+        let mut results = vec![];
+        for (k, v) in self.programs.iter_mut() {
+            if v.kind() == program_type && v.if_index() == if_index && v.direction() == direction {
+                results.push((k, v))
+            }
+        }
+        results
+    }
+
+    fn sort_programs(
+        &mut self,
+        program_type: ProgramType,
+        if_index: Option<u32>,
+        direction: Option<Direction>,
+    ) {
+        let mut extensions = self
+            .programs
+            .iter_mut()
+            .filter_map(|(k, v)| {
+                if v.kind() == program_type {
+                    if v.if_index() == if_index && v.direction() == direction {
+                        Some((k, v))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<(&Uuid, &mut Program)>>();
+
+        extensions.sort_by_key(|(_, b)| {
+            (
+                b.priority(),
+                b.attached().unwrap(),
+                b.data()
+                    .expect("All Bpfd programs should have ProgramData")
+                    .name()
+                    .to_owned(),
+            )
+        });
+        for (i, (_, v)) in extensions.iter_mut().enumerate() {
+            v.set_position(Some(i));
+        }
+    }
+
+    fn kernel_infos(&self) -> HashMap<u32, Program> {
+        self.programs
+            .values()
+            .map(|p| {
+                let kernel_info = p
+                    .data()
+                    .expect("All Bpfd programs should have ProgramData")
+                    .kernel_info()
+                    .expect("Loaded Bpfd programs should have kernel information");
+
+                (kernel_info.id, p.to_owned())
+            })
+            .collect()
+    }
+}
+
+pub(crate) struct BpfDispatchers {
+    dispatchers: HashMap<DispatcherId, Dispatcher>,
+}
+
+impl BpfDispatchers {
+    pub(crate) fn remove(&mut self, id: &DispatcherId) -> Option<Dispatcher> {
+        self.dispatchers.remove(id)
+    }
+
+    pub(crate) fn insert(&mut self, id: &DispatcherId, dis: Dispatcher) -> Option<Dispatcher> {
+        let id = id.clone();
+        self.dispatchers.insert(id, dis)
+    }
+}
+
 impl BpfManager {
     pub(crate) fn new(config: Config, commands: mpsc::Receiver<Command>) -> Self {
         Self {
             config,
-            dispatchers: HashMap::new(),
-            programs: HashMap::new(),
+            dispatchers: BpfDispatchers {
```

Comment:
Since you've gone through the bother of making them their own types, add a `new()` function 😉

Reply:
Lol done